LFS-1073: Refactor the vocabulary question to use MultipleChoice component

OrdiNeu opened the pull request:
This should also fix:
* [LFS-947](https://phenotips.atlassian.net/browse/LFS-947): Vocabulary questions where maxAnswers = 1 present unexpected behavior
* [LFS-353](https://phenotips.atlassian.net/browse/LFS-353): Remove radio button when field is defined as Data Type = vocabulary & MinAnswers=MaxAnswers=1
* [LFS-1046](https://phenotips.atlassian.net/browse/LFS-1046): Vocabulary questions with maxAnswers=1 cannot be saved
* [LFS-1082](https://phenotips.atlassian.net/browse/LFS-1082): Vocabulary questions allow selecting more than the maximum number of allowed answers

**Test branch**:  [`LFS-1073-test`](https://github.com/ccmbioinfo/lfs/tree/LFS-1073-test) which contains one questionnaire named `Demo` with 4 vocabulary questions with different settings, all having HANCESTRO as source vocabulary.

OrdiNeu committed:
[wip] LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

Adding stuff now works.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/VocabularyQuestion.jsx:
```diff
@@ -80,14 +81,18 @@ function VocabularyQuestion(props) {
       {...props}
       >
-      <VocabularySelector
+      <MultipleChoice
```
`MultipleChoice` now handles `AnswerInstructions`, so line 77 (which initializes `currentAnswers`) can now be removed and line 81 should be replaced with `disableInstructions`. Otherwise we can get duplicate `AnswerInstructions`, like in this test question: 

![image](https://user-images.githubusercontent.com/651980/113936900-73613000-97c6-11eb-958a-970da1415041.png)

marta- replied:
Also for some reason the default answer options are not populated with the right labels. In Demographics, we don't actually use vocab ids yet, but I created branch `LFS-1073-test` with a Demo questionnaire with 4 questions hooked to HANCESTRO. This is what one of the questions looks like:

![image](https://user-images.githubusercontent.com/651980/113939923-9e4d8300-97ca-11eb-8775-25ed1b879b41.png)

OrdiNeu replied:
Should be resolved in `76822de` now :heavy_check_mark:

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/VocabularyQuestion.jsx:
```diff
@@ -80,14 +81,18 @@ function VocabularyQuestion(props) {
+        customInputProps = {{
+          vocabularies: sourceVocabularies
+        }}
         vocabularyFilter = {vocabularyFilter}
```
`vocabularyFilters` should be part of `customInputProps`?

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/VocabularyQuestion.jsx:
```diff
@@ -80,14 +81,18 @@ function VocabularyQuestion(props) {
+          vocabularies: sourceVocabularies
+        }}
         vocabularyFilter = {vocabularyFilter}
         max = {maxAnswers}
```
max is not a valid property for MultipleChoice

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/VocabularyQuestion.jsx:
```diff
@@ -80,14 +81,18 @@ function VocabularyQuestion(props) {
+          vocabularies: sourceVocabularies
+        }}
         vocabularyFilter = {vocabularyFilter}
         max = {maxAnswers}
         defaultSuggestions = {defaultSuggestions}
         source = {sourceVocabularies}
         selectionUpdated = {(count) => setCurrentAnswers(count)}
```
I believe all of these can be removed

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/MultipleChoice.jsx:
```diff
@@ -297,35 +300,50 @@ function MultipleChoice(props) {
+  let ghostInput = (input || textbox || customInput) && (<div className={isBare ? classes.bareAnswer : classes.searchWrapper}>
+      {
+        customInput ?
+          <CustomInput
```
Bug: For questions with maxAnswers=1, the default multiple choice input retains the entered value, but the vocab input clears it after selection. The value is selected and saved (as can be verified in the JSON or in view mode/patient chart) but the user never sees it in edit mode.

OrdiNeu replied:
Fixed in `aa203c6` :heavy_check_mark:

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

Fix an issue where the labels for entered questions were not being
applied correctly.

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

Fix a bug that occurs when maxAnswers=1

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

Fix an issue when reloading the page with maxAnswers=1

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

Remove unused code.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/VocabularyQuestion.jsx:
```diff
@@ -63,31 +43,21 @@ import VocabularySelector from "../vocabSelector/select.jsx";
 
   return (
     <Question
-      currentAnswers={currentAnswers}
```
Forgot to add  the `disableInstructions` prop here instead of passing `currentAnswers`. Now we still display 2 sets of "instructions", but only the second one actually takes existing answrs into account.

![image](https://user-images.githubusercontent.com/651980/114086191-55f09c80-9880-11eb-80a3-e4175bf6eaf5.png)

OrdiNeu replied:
Fixed in `ad42edd` and `75e6993` :heavy_check_mark:

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

Replace the Vocabularies prop with sourceVocabularies in the
questionDefinition of query.jsx

marta- reviewed (changes requested):
Bug: The input is no longer cleared after selecting a suggested entry for vocab questions with maxAnswers != 1. Eg: type i, select "Italian" -> the option is added, the suggestion list disappears, but the input still has "i" as the value.

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

Remove the double answer hint

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
componento

Clear the input after selection

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/VocabularyQuestion.jsx:
```diff
@@ -63,31 +43,21 @@ import VocabularySelector from "../vocabSelector/select.jsx";
-    .reduce((object, value) => ({...object, [value[VALUE_POS]]: value[LABEL_POS]}), {});
-
-  let [currentAnswers, setCurrentAnswers] = useState(Array.of(props.existingAnswer?.[VALUE_POS]?.value || []).flat().length);
+  let { questionDefinition, sourceVocabularies } = { ...props.questionDefinition, ...props };
```
Unless I'm missing something, `sourceVocabularies` doesn't appear to be used here directly

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

Apply code review improvement by @marta-

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/query.jsx:
```diff
@@ -84,7 +85,7 @@ class VocabularyQuery extends React.Component {
 
   render() {
-    const { classes, defaultValue, disabled, inputRef, noMargin, isNested, onInputFocus, placeholder, searchDefault, vocabularies } = this.props;
+    const { classes, defaultValue, disabled, inputRef, noMargin, isNested, onInputFocus, placeholder, searchDefault, value, vocabularies } = this.props;
```
`vocabularies` appears to have been replaced by `questionDefinition.sourceVocabularies`.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/VocabularyQuestion.jsx:
```diff
@@ -43,7 +43,7 @@ import VocabularyQuery from "../vocabQuery/query.jsx";
 function VocabularyQuestion(props) {
   let { classes, ...rest } = props;
-  let { questionDefinition, sourceVocabularies } = { ...props.questionDefinition, ...props };
+  let { questionDefinition } = { ...props.questionDefinition, ...props };
```
Shouldn't this just be 
```
let { questionDefinition } = props;
```
?

OrdiNeu replied:
Fixed in `fb92521`, I've also gotten rid of the unused `classes` and `...rest` variables.

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

Code cleanup suggestions by @marta-

sashaandjic commented:
Found this in the test branch:
![image](https://user-images.githubusercontent.com/23243244/114250138-b3f9ae80-996a-11eb-8b14-56bd8ad418e2.png)

sashaandjic commented:
- set maxAnswer = 1 for Demographic form's Race/Ethnicity filed
- create Demographic form and select a vocabulary entry that is not predefined ( e.g. St. Helenian)
- now select one of the predefined options
- St. Helenian is still there (bug, it should be removed)
- save, close and reopen the form -> St. Helenain is no longer there (correct)

sashaandjic commented:
- set maxAnswer = 1 for Demographic form's Race/Ethnicity field
- create Demographic form and select an option from the vocabulary  (not the predefined one, e.g.  Haitian)
- save and close the form
- Haitian option is shown in the patient chart (correct)
- open the form -> Haitian option is there (c0rrect)
- click on Edit button 
- Haitian is gone (bug)
- close the form -> Haitian option is shown in the patient chart (correct)

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

Fix an issue where the answer would not pre-load in radio answers.

OrdiNeu commented:
@sashaandjic 
> Bug in LFS-1073 branch:
I've fixed the third bug, and I think the second bug is gone too in `813d457`? I'm not sure that i followed your replication correctly though, since pressing enter didn't select `Haitian` for me.

sashaandjic commented:
Problems are still present. 
- set maxAnswer = 1 for Demographic form's Race/Ethnicity field
- create Demographic form, type and select British
- save and close the form
- British is shown in the patient chart (correct)
- open the form -> British option is there (correct)
- click on the Edit button
- British is gone (bug)
close the form -> British option is shown in the patient chart (correct)
set maxAnswer = 1 for Demographic form's Race/Ethnicity field
create Demographic form and select an option from the vocabulary (not the predefined one, e.g. Haitian)
save and close the form
Haitian option is shown in the patient chart (correct)
open the form -> Haitian option is there (correct)
click on the Edit button
Haitian is gone (bug)
close the form -> Haitian option is shown in the patient chart (correct)
![image](https://user-images.githubusercontent.com/23243244/114476073-937a6000-9bc7-11eb-9f8d-89c09955c335.png)

sashaandjic commented:
Another issue:
- set maxAnswer = 2
- user can select more than 2 when browsing vocabulary, without form displaying some kind of warning
![image](https://user-images.githubusercontent.com/23243244/114477036-80688f80-9bc9-11eb-8359-fae09819ab7d.png)
- after saving and closing only 2 are saved, but user may not predict which two
![image](https://user-images.githubusercontent.com/23243244/114477084-9aa26d80-9bc9-11eb-9df0-1e3a9e5a81f2.png)

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

Ensure that the query is disabled when everything else is

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

Fix an issue where the user-defined input would not be used in the input
box for radio/bare MultipleChoice questions.

OrdiNeu commented:
@sashaandjic 
> Problems are still present. 

I think I fixed it in `f76f93d`, please confirm.

sashaandjic commented:
With maxAnswers = 1, problem when clicking on `ENTER `upon vocabulary selection is still there.
Also new proble detected:
- set maxAnswers = 1 for Race/Ethnicity reported 
- select any predefined option
- save, close, reopen
- click on Edit button --> the selection is copied on the search line
![image](https://user-images.githubusercontent.com/23243244/114618501-6b4a3a00-9c77-11eb-8711-07c3c6ebd9ca.png)

marta- committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice component
Fix styling after rebase on dev

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component
Fix an issue where selected pre-defined options appeared twice

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/MultipleChoice.jsx:
```diff
@@ -91,9 +92,12 @@ function MultipleChoice(props) {
-  const ghostSelected = selection.some(element => {return element[VALUE_POS] === GHOST_SENTINEL;});
+
+  // If this is a bare input or radio input, we need to pre-populate the blank input with the custom answer (if available)
+  let radioPrefill = (isBare || (isRadio && defaults.indexOf(initialSelection[0]) < 0)) && existingAnswer?.[1] || '';
```
Rename as `inputPrefill` to avoid confusion (it's also done for `isBare`, not just for `isRadio`.

OrdiNeu replied:
Fixed in `274648c` :heavy_check_mark:

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

Fix an issue where loaded options were not being saved properly
Fix an issue where vocab options were incorrectly being discounted
from the number of selected options.

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

Fix various issues caused by not updating the ghostName on vocab
multiple choice questions
Rename radioPrefill to inputPrefill as suggested by @marta-

sashaandjic commented:
Installed HANCESTRO and tried to create Demographic form, then got this:
![image](https://user-images.githubusercontent.com/23243244/114801969-d163b980-9d6a-11eb-853c-bb56119ce76f.png)

marta- committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice component

Fix JS console error when there's no initial value

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

Fix the behaviour when selecting a default option in a radio button

sashaandjic commented:
set maxAnswers = 1
- type something and pick an entry from the vocabulary
- hit enter few times
![image](https://user-images.githubusercontent.com/23243244/114924255-c9078f00-9dfb-11eb-8e34-42539a4759ec.png)

marta- commented:
> set maxAnswers = 1
> 
> * type something on the search barand pick an entry from the vocabulary
> * hit enter few times

Not a show stopper, and the behavior appears similar to what happens in dev, maybe even slightly better. To be reported and approached separately.

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

Fix an issue where pressing enter creates newlines in the query

OrdiNeu commented:
@sashaandjic 
> set maxAnswers = 1
> 
>     * type something on the search barand pick an entry from the vocabulary
> 
>     * hit enter few times

Fixed in `f9b8561` :heavy_check_mark:

sashaandjic commented:
Confirmed in dev branch
> * [LFS-1045](https://phenotips.atlassian.net/browse/LFS-1045): displayValues for Vocabulary questions are inconsistently exported as single string or array (TO BE CONFIRMED)

sashaandjic commented:
Set minAnswers=maxAnswers=1 for Demographic form's `Co-morbidites` field:
![image](https://user-images.githubusercontent.com/23243244/114933485-b777b480-9e06-11eb-9e8c-b1d0d2be48ae.png)

Create a form and enter the value. All is good:
![image](https://user-images.githubusercontent.com/23243244/114933543-c8282a80-9e06-11eb-9ec8-b4190c945f62.png)

Now delete the entry and try to select another value --> the focus is gone. The only way around is to save, close and reopen the form.

sashaandjic commented:
- set maxAnswers = 1 for Race/Ethnicity field
- type and select Mauritanian
- save, close, open and edit the form
- select any predefine option
- the "Maurutanian" is not cleared
![image](https://user-images.githubusercontent.com/23243244/114934264-b1360800-9e07-11eb-9cbf-53cab1c2c16e.png)

sashaandjic commented:
[LFS-1045](https://phenotips.atlassian.net/browse/LFS-1045) is not fixed.

![image](https://user-images.githubusercontent.com/23243244/114934926-8ef0ba00-9e08-11eb-9d21-893377b7b570.png)

![image](https://user-images.githubusercontent.com/23243244/114934804-636dcf80-9e08-11eb-9100-5ce169201854.png)

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

Fix bugs in multiple choice with custom inputs on non-vocab questions.

marta- commented:
> LFS-1045 is not fixed.

Yep, 1045 looks completely unrelated, I'll remove ot from the "related" list.

OrdiNeu commented:
@sashaandjic 
>     * the "Maurutanian" is not cleared

Is this unexpected behaviour for all MultipleChoice questions? Currently this is how all multiple choice questions with `maxAnswers=1` and an input field work.

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

Fix an issue that can occur after selecting a term, typing a new term,
and becoming unable to finish typing.

OrdiNeu commented:
@sashaandjic 
> Also modifying the existing selection is impossible without deleting it first.
Should be fixed in  `454779c` :heavy_check_mark:

marta- commented:
> @sashaandjic
> 
> > ```
> > * the "Maurutanian" is not cleared
> > ```
> 
> Is this unexpected behaviour for all MultipleChoice questions? Currently this is how all multiple choice questions with `maxAnswers=1` and an input field work.

Actually you're right, we should keep this. Accidentally clicks another (predefined) radio button and they see the former value vanish isn't a great user experience.

The only thing I would fix, if easy, is the label overlapping the entered value. But it's not a show stopper.

Everything else looks good to me so far otherwise.

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/query.jsx:
```diff
@@ -128,6 +134,7 @@ class VocabularyQuery extends React.Component {
       )}
       defaultValue={defaultValue}
       placeholder={placeholder}
+      value={value || this.state.inputValue}
```
We should not be using both `value` and `defaultValue`. 

According to[ Mui documentation](https://material-ui.com/api/input/):

> `defaultValue` | The default `input` element value. Use when the component is not controlled.
> `value` | The value of the `input` element, required for a controlled component.

As expected, this warning popped up in the console:

```
Warning: A component is changing an uncontrolled input of type text to be controlled. Input elements should not switch from uncontrolled to controlled (or vice versa). Decide between using a controlled or uncontrolled input element for the lifetime of the component
```

OrdiNeu replied:
Fixed in `8502ed1` :heavy_check_mark:

marta- commented:
> The only thing I would fix, if easy, is the label overlapping the entered value. But it's not a show stopper.

@OrdiNeu in #526 (PR for 947, now closed without merging since this PR makes it obsolete), I hid the label when there's a value inside to "fix" (hide) this issue. This may be a decent compromise if there isn't a more elegant way.

https://github.com/ccmbioinfo/lfs/pull/526/files#diff-d3aef3d6d518c380ce9d9e85ab29f65db164b6a812cf571cdd83dc6c9930020bL146-R150

marta- reviewed (commented):
Minor UI/UX improvement suggestions:
* the [magnifier glass InputAdornment](https://github.com/ccmbioinfo/lfs/pull/589/files#diff-d3aef3d6d518c380ce9d9e85ab29f65db164b6a812cf571cdd83dc6c9930020bR131-R133) acts like a button (clicking it results in relaunching the search) but its cursor is misleading. It should have `cursor: pointer`.
* the [search wrapper's padding top](https://github.com/ccmbioinfo/lfs/pull/589/files#diff-8925e54353bd3fe68d03daf91997cb83d41dd9bdb63d193c7b646b599e20a078R144) makes the input misaligned with the radio button, i think it should be removed

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

Fix an issue with the input suggestion being displayed even after a
value was entered
Rename searchDefault with label

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

Apply UI improvements suggested by @marta-

OrdiNeu commented:
@marta- 
> @OrdiNeu in #526 (PR for 947, now closed without merging since this PR makes it obsolete), I hid the label when there's a value inside to "fix" (hide) this issue. This may be a decent compromise if there isn't a more elegant way. See lines 146-150 in query.jsx: https://github.com/ccmbioinfo/lfs/pull/526/files#diff-d3aef3d6d518c380ce9d9e85ab29f65db164b6a812cf571cdd83dc6c9930020bL146-R150
> 
> (Note: `searchDefault` is also renamed to `label`, which IMO is a better fitting name)

Fixed in `74b0f81` :heavy_check_mark: 

> Minor UI/UX improvement suggestions:
Applied in `8502ed1` :heavy_check_mark:

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

Fix an issue with vocab filters that reopens the dialog after selecting
an option

sashaandjic reviewed (approved):
tested
- [x] LFS-947: Vocabulary questions where maxAnswers = 1 present unexpected behavior
- [x] LFS-353: Remove radio button when field is defined as Data Type = vocabulary & MinAnswers=MaxAnswers=1
- [x] LFS-1046: Vocabulary questions with maxAnswers=1 cannot be saved
- [x] LFS-1082: Vocabulary questions allow selecting more than the maximum number of allowed answers

sashaandjic reviewed (approved):
re-tested

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/VocabularyQuestion.jsx:
```diff
@@ -27,67 +27,36 @@ import Question from "./Question";
-// <VocabularyQuestion
 //   questionDefintion={{
 //     text: "Does the patient have any skin conditions?",
 //     sourceVocabularies={["hpo"]},
 //     vocabularyFilter: ["HP:0000951"]
```
This should be updated soon to match the new reality.

OrdiNeu replied:
This should be the new reality -- I cut out all of the props that did not use `questionDefinition`.

sdumitriu replied:
Not quite, besides the typo in `questionDefintion`, `sourceVocabularies` is just an array, not an array in an object (which is invalid syntax), while `vocabularyFilter` is plural and actually an object.

sdumitriu reviewed (changes requested):
Bug 1:

- install HANCESTRO
- modify Demographics / Race to have maxAnswers=1
- create a new form
- search for "Danish" and select it
- save, close form, reopen in edit mode -> no value is selected, although it is still selected in the background, and even saving the form will not lose the value

Bug 2:
- on the previous form, select one of the predefined options, like Indian, save, close the form
- modify the questionnaire to remove all the predefined options
- reopen the form -> no value displayed, yet the input is disabled

In other words: when maxAnswers=1, hijinks ensue.

marta- commented:
> In other words: when maxAnswers=1, hijinks ensue.

Before this PR, vocabulary questions with maxAnswers = 1 were completely non-functional (see[ LFS-947](https://phenotips.atlassian.net/browse/LFS-947)). These bugs are actually an improvement :)

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

Disable going back to the input if maxAnswers == 1, and also refactor a
Vocab-specific prop out of MultipleChoice

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/VocabularyQuestion.jsx:
```diff
@@ -27,67 +27,36 @@ import Question from "./Question";
-//   />
-// <!-- Alternate method of specifying arguments -->
-// <VocabularyQuestion
 //   questionDefintion={{
 //     text: "Does the patient have any skin conditions?",
 //     sourceVocabularies={["hpo"]},
 //     vocabularyFilter: ["HP:0000951"]
```
```suggestion
//   questionDefinition={{
//     text: "Does the patient have any skin conditions?",
//     sourceVocabularies=["HP"],
//     vocabularyFilters: {"HP": ["HP:0000951"]}
```

OrdiNeu replied:
Fixed in `0797abd` :heavy_check_mark:

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

Fix a bug where clicking on an empty maxAnswers==1 multiple choice would
select 'custom-input'

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

Fix the ghost value persisting after removal.

OrdiNeu committed:
LFS-1073: Refactor the vocabulary question to use MultipleChoice
component

Fix the docstring for VocabularyAnswer as suggested by @sdumitriu

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/VocabularyQuestion.jsx:
```diff
@@ -27,67 +27,41 @@ import Question from "./Question";
 //     text: "Does the patient have any skin conditions?",
-//     sourceVocabularies={["hpo"]},
-//     vocabularyFilter: ["HP:0000951"]
+//     sourceVocabularies=["HP"],
```
Oh, so many errors in here...

```suggestion
//     sourceVocabularies: ["HP"],
```

sdumitriu committed:
[misc] Another typo fix

sashaandjic reviewed (commented):
re-tested
